### DISCUSSION TITLE
Can't remove images from a bucket with a trigger

### DISCUSSION THREAD

**Discussion by shalawfatah** [supabase/supabase#13912] [Questions]
I have this table called `products` which has this column called `images` type of `jsonb[]`, now I try to remove all images associated with a product when a product is removed using this function and trigger in SQL,

```
CREATE OR REPLACE FUNCTION delete_product_images()
RETURNS trigger AS $$
DECLARE
  image_links TEXT[];
  link TEXT;
BEGIN
  -- Retrieve image links from deleted row's JSON data
  SELECT json_array_elements(old.images) INTO image_links;
  
  -- Delete each image from the product-images bucket
  FOREACH link IN ARRAY image_links LOOP
    PERFORM supabase.storage.object_delete('product-images', link::text);
  END LOOP;

  RETURN NULL;
END;
$$ LANGUAGE plpgsql;

CREATE TRIGGER delete_product_trigger
AFTER DELETE ON products
FOR EACH ROW
EXECUTE FUNCTION delete_product_images();
```

However, when I try to delete, I get this error:
```
Failed to run sql query: function json_array_elements(jsonb[]) does not exist
```

I have tried type casting: ` SELECT json_array_elements(old.images::jsonb and ::jsonb[]) INTO image_links;` but none of it works. How to solve this problem? I can't manually remove images from the database.

**Comment by GaryAustin1** [comment]
It  is jsonb_array_elements for jsonb...

Also are you in your delete function doing http calls to the storage-api?  You can't just delete rows in storage.objects.

**Comment by StevePhuc** [comment]
https://supabase.com/docs/guides/getting-started/tutorials/with-expo-react-native#storage-management 

> To automatically remove obsolete storage objects, extend the database
triggers. Note that it is not sufficient to delete the objects from the
storage.objects table because that would orphan and leak the actual storage objects in
the S3 backend. Instead, invoke the storage API within Postgres via the http extension.
